rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,159 @@
+
+### Index Management
+
+For every document that needs to be processed for an index, we have to run the document through the javascript query server to get the emitted keys and values. This means that it won’t be possible to update a map/reduce index in the same transaction that a document is updated. To account for this, we will need to keep an `id index` similar to the `id tree`  that is currently keep in CouchDB. This index will hold the document id as the key and the value would be the keys that were emitted. CouchDB will use this information to know which fields need to be updated, added or removed from the index when a document is changed.  A data model for this would be:
+
+{?DATABASE, ?VIEWS, <view_signature>, ?VIEWS, ?ID_INDEX, <_id>, <view_id>} -> [emitted keys]
+
+Each index will be built and updated via the Background job queue [RFC Link TBD]. When a request for a view is received, the request process will add a job item onto the background queue for the index to be updated. A worker will take the item off the queue and update the index. Once the index has been built, the request will return with the results. This process can also be optimised in two ways. Firstly, using a new couch_events system to listen for document changes in an database and then adding indexing jobs to the queue to keep indexes warm. The second optimisation is if the index only requires a small update, rather update the index in the http request process instead of doing the work via the background queue.
+
+Initially the building of an index will be a single worker running through the changes feed and creating the index. Ideally it would be nice to parallelise that work so that multiple workers could build the index at the same time. This will reduce build times. This can be done by fetching the boundary keys for the changes feed, splitting those key ranges amongst different workers to build different parts of the index. This will require that for each document update processed, the worker must check the revision key space to determine if the document is the latest revision of the document. If it is not it should discard it. The other requirement is that CouchDB could only start serving the index once the update sequence is up to date AND all the workers have completed building their section of the index.
+
+### View clean up
+
+When a design document is changed, new indexes will be built and grouped under a new  `View Signature` . The old map indexes will still be in FDB.  CouchDB will need to run a clean up process that scans all the `View Signature`s and links them to a design document. If a `View Signature` is not linked to any design doc, that `View Signature` namespace can be removed. This will be done via the background jobs scheduler.
```

Comment:
As written this would be fairly inefficient as we'd have to scan all of the index content to get the list of view signatures that currently exist since there's no concise listing.

A simple approach to avoid that can also remove the need for the background cleaning process completely. If we were to store an pair of indexes for view signatures of the form:

`(<database>, ?VIEW_DDOC_IDS, DesignDocId) = ViewSignature`
`(<database>, ?VIEW_SIGS, ViewSignature) = Counter`

When a design document is created, we store the view signature and bump the view signature counter. One update or deletion we'd get the old signature and decrement the counter. And on 0 we'd clear out the old view index.

Reply:
Nice idea. I've added it to the rfc.